src/diagnostic-channel-publishers/src/mongodb.pub.ts

```diff
@@ -54,11 +54,54 @@ const mongodbPatchFunction: PatchFunction = function(originalMongo) {
     return originalMongo;
 };
 
+const mongodb3PatchFunction: PatchFunction = function(originalMongo) {
+    const listener = originalMongo.instrument({
+        operationIdGenerator: {
+            next: function() {
+                return channel.bindToContext((cb) => cb());
+            },
+        },
+    });
+    const eventMap = {};
+    listener.on("started", function(event) {
+        if (eventMap[event.requestId]) {
+            // Note: Mongo can generate 2 completely separate requests
+            // which share the same requestId, if a certain race condition is triggered.
+            // For now, we accept that this can happen and potentially miss or mislabel some events.
+            return;
+        }
+        eventMap[event.requestId] = event;
+    });
+
+    listener.on("succeeded", function(event) {
+        const startedData = eventMap[event.requestId];
+        if (startedData) {
+            delete eventMap[event.requestId];
+        }
+        channel.publish<IMongoData>("mongodb", {startedData, event, succeeded: true});
+    });
+
+    listener.on("failed", function(event) {
+        const startedData = eventMap[event.requestId];
+        if (startedData) {
+            delete eventMap[event.requestId];
+        }
+        channel.publish<IMongoData>("mongodb", {startedData, event, succeeded: false});
+    });
+
+    return originalMongo;
+};
+
 export const mongo2: IModulePatcher = {
-    versionSpecifier: ">= 2.0.0 <= 2.2.x",
+    versionSpecifier: ">= 2.0.0 <= 3.0.5",
     patch: mongodbPatchFunction,
 };
+export const mongo3: IModulePatcher = {
+    versionSpecifier: "> 3.0.5",

```

Comment:
We need to mark a last supported version. Otherwise we can start breaking apps in the future.

Reply:
I originally left it blank because mongodb provides an API for monitoring tools to hook onto the library, so we are not directly tied to any breaking changes in the API. They say it will [be forward compatible](https://github.com/mongodb/node-mongodb-native/blob/7298c76e2a1b1b2e37a211860f35f84fad5a4c3b/docs/reference/content/reference/management/apm.md#apm), but our usage of it was broken at `3.0.6`. Limiting to < 4 might suffice, but we are exposed to an indirect API that may "break" on non major version bumps.